src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs

```diff
@@ -0,0 +1,529 @@
+            _cts.Token.Register(AbortRequest);
+
+            // Wrap the cancellationToken in a using so that it can be disposed of whether
+            // we successfully connected or failed trying.
+            // Otherwise any timeout/cancellation would apply to the full session.
+            // In the failure case we need to release the references and dispose of the objects.
+            using (cancellationToken.Register(() => tcsConnect.TrySetCanceled()))
+            {
+                try
+                {
+                    JavaScript.Array? subProtocols = null;
+                    if (requestedSubProtocols?.Count > 0)
+                    {
+                        subProtocols = new JavaScript.Array();
+                        foreach (var item in requestedSubProtocols)
+                        {
+                            subProtocols.Push(item);
+                        }
+                    }
+                    _innerWebSocket = subProtocols == null ? new HostObject("WebSocket", uri.ToString()) : new HostObject("WebSocket", uri.ToString(), subProtocols);
+                    _innerWebSocket.SetObjectProperty("binaryType", "arraybuffer");
+
+                    subProtocols?.Dispose();
+
+                    // Setup the onError callback
+                    _onError = new Action<JSObject>((errorEvt) =>
+                    {
+                        errorEvt.Dispose();
+                    });
+
+                    // Attach the onError callback
+                    _innerWebSocket.SetObjectProperty("onerror", _onError);
+
+                    // Setup the onClose callback
+                    _onClose = new Action<JSObject>((closeEvt) =>
+                    {
+                        _innerWebSocketCloseStatus = (WebSocketCloseStatus)closeEvt.GetObjectProperty("code");
+                        _innerWebSocketCloseStatusDescription = closeEvt.GetObjectProperty("reason")?.ToString();
+                        var mess = new ReceivePayload(ArraySegment<byte>.Empty, WebSocketMessageType.Close);
+                        _receiveMessageQueue.BufferPayload(mess);
+
+                        if (!tcsConnect.Task.IsCanceled && !tcsConnect.Task.IsCompleted && !tcsConnect.Task.IsFaulted)
+                        {
+                            tcsConnect.SetException(new WebSocketException(WebSocketError.NativeError));
+                        }
+                        else
+                        {
+                            _tcsClose?.SetResult(true);
+                        }
+
+                        closeEvt.Dispose();
+                    });
+
+                    // Attach the onClose callback
+                    _innerWebSocket.SetObjectProperty("onclose", _onClose);
+
+                    // Setup the onOpen callback
+                    _onOpen = new Action<JSObject>((evt) =>
+                    {
+                        using (evt)
```

Comment:
Why is the using needed ?

Reply:
Really it is not needed anymore with the reference counting PR but since it is a native JSObject I prefer to release it as soon as possible.

I can change this if it is preferred.